src/config.rs

```diff
@@ -0,0 +1,195 @@
+pub mod settings;
+pub mod storage;
+
+use crate::config::settings::{Setting, SettingInfo};
+use serde_derive::Deserialize;
+use serde_json::Value;
+use std::collections::HashMap;
+use std::mem;
+use wildmatch::WildMatch;
+
+const SETTINGS_JSON: &str = include_str!("./config/settings.json");
+
+#[derive(Debug, Deserialize)]
+struct JsonEntry {
+    key: String,
+    #[serde(rename = "type")]
+    entry_type: String,
+    default: String,
+    description: String,
+}
+
+/// StorageAdapter is the interface for storing and retrieving settings
+/// This can be used to store settings in a database, json file, etc
+pub trait StorageAdapter {
+    /// Retrieves a setting from the storage
+    fn get_setting(&self, key: &str) -> Option<Setting>;
+    /// Stores a given setting to the storage
+    fn set_setting(&mut self, key: &str, value: Setting);
+    /// Retrieves all the settings in the storage in one go. This is used for preloading the settings
+    /// into the ConfigStore and is more performant normally than calling get_setting manually for each
+    /// setting.
+    fn get_all_settings(&self) -> HashMap<String, Setting>;
+}
+
+/// Configuration store is the place where the gosub engine can find all configurable options
+pub struct ConfigStore {
+    /// A hashmap of all settings so we can search o(1) time
+    settings: HashMap<String, Setting>,
+    /// A hashmap of all setting descriptions, default values and type information
+    settings_info: HashMap<String, SettingInfo>,
+    /// Keys of all settings so we can iterate keys easily
+    setting_keys: Vec<String>,
+    /// The storage adapter used for persisting and loading keys
+    storage: Box<dyn StorageAdapter>,
+}
+
+impl ConfigStore {
+    /// Creates a new store with the given storage adapter and preloads the store if needed
+    pub fn new(storage: Box<dyn StorageAdapter>, preload: bool) -> Self {
```

Comment:
I suspect we can avoid using `Box<dyn ...>` if we replace it with a generic struct parameter.  But in this specific case I doubt it would make a difference, since this is not (presumably) critical path code.